### PR TITLE
 AST: Exclude noncanonical params from `GenericSignature::enumeratePairedRequirements`.

### PR DESCRIPTION
**Pull request by jckarter**
Fixes a bug where having two generic params same-typed together would lead to crashes in IRGen, and maybe other places, that only expect to deal with canonical params.

### CODE REVIEW

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by jckarter** [comment]
@swift-ci Please test source compatibility

File: lib/AST/GenericSignature.cpp [hunk 1]
```diff
@@ -345,14 +345,26 @@ bool GenericSignature::enumeratePairedRequirements(
...
+      // parameter itself.
+      if (req.getSecondType()->isTypeParameter()) continue;
+      
+      // Otherwise, the generic parameter is concrete.
+      gp = req.getFirstType()->getAs<GenericTypeParamType>();
+      if (!gp) continue;
+    }
 
     unsigned index = GenericParamKey(gp).findIndexIn(genericParams);
     genericParamsAreConcrete[index] = true;
```
Comment (DougGregor) [inline lib/AST/GenericSignature.cpp]: `genericParamsAreConcrete` is now a misnomer; `genericParamIsNonCanonical`, perhaps?

**Review by DougGregor — APPROVED** [review]

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by jckarter** [comment]
@swift-ci Please test source compatibility

**Pull request merged**